### PR TITLE
rgw multisite: remove sharding from bucket full sync

### PR DESCRIPTION
**Pull request by cbodley**
currently, bucket full sync runs once for each bucket index shard, by listing all objects in that shard and fetching them. this coupling with the index shards is problematic for resharding, because any full syncs in progress would be unable to continue their listing after a reshard

by moving full sync from per-bucket-shard to per-bucket, we can use a normal ordered bucket listing, which is stable across reshards. to track this full sync status, we add a new per-bucket sync status object, along with some backward-compatibility logic to skip a full sync if all of its shard status objects exist and show incremental sync

when starting full sync, a cls lock is acquired on this full sync status object to prevent other data sync shards from duplicating the work. the shards that fail to get this lock will retry until the full sync finishes

in the future, this bucket sync status object will be used to coordinate the per-shard incremental sync across reshard events

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by cbodley** [comment]
jenkins test signed

**Comment by cbodley** [comment]
https://pulpito.ceph.com/cbodley-2020-10-08_17:50:30-rgw-wip-cbodley-testing-distro-basic-smithi/

**Comment by cbodley** [comment]
> i seem to have broken the 'bucket sync disable' tests for pubsub again

addressed by using exclusive create/cls_version in `InitBucketShardStatusCR` to avoid overwriting the incremental marker in shard status. we have to save that marker while sync is disabled for pubsub zones, because they can't restart with a full sync; and if they restart incremental sync, they'll sync back up to the SYNCSTOP entry and get stuck in a loop

**Comment by yehudasa** [comment]
Should we even continue supporting `bucket sync disable`? The assumption that a bucket syncs symmetrically doesn't align with bucket granularity sync. Maybe the way to do it is through creating a sync policy that forbids that pipe?

**Comment by cbodley** [comment]
@yehudasa remember the sync policy stuff is still experimental, and has to be manually enabled in the zonegroup configuration once all zones are upgraded to octopus+. once that's set up, we'll either need some automated conversion logic for disabled buckets, or documentation on how to set up the equivalent policy. until then, i think we're stuck supporting `bucket sync disable`

**Comment by cbodley** [comment]
pushed an update which seems to be passing multisite tests, except for those with `@attr('bucket_sync_disable')` which we'll revisit later

File: qa/suites/rgw/multisite/tasks/test_multi.yaml [hunk 1]
```diff
@@ -24,3 +24,4 @@ tasks:
 - rgw-multisite-tests:
     config:
       reconfigure_delay: 60
+    args: ['-a', '!bucket_sync_disable'] # filter out disable/enable tests
```
Comment (cbodley) [inline qa/suites/rgw/multisite/tasks/test_multi.yaml:27]: strange, this isn't working even though `nosetests -a '!bucket_sync_disable'` works fine on the command line:
```
2020-11-12T22:33:05.871 INFO:tasks.rgw_multisite_tests:running rgw multisite tests on 'tasks.rgw_multi.tests' with args=['-a', '!bucket_sync_disable']
2020-11-12T22:33:05.873 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
  File "/usr/lib/python3.6/optparse.py", line 1387, in parse_args
    stop = self._process_args(largs, rargs, values)
  File "/usr/lib/python3.6/optparse.py", line 1431, in _process_args
    self._process_short_opts(rargs, values)
  File "/usr/lib/python3.6/optparse.py", line 1513, in _process_short_opts
    raise BadOptionError(opt)
optparse.BadOptionError: no such option: -a
```
https://nose.readthedocs.io/en/latest/usage.html#options

Comment (cbodley) [reply qa/suites/rgw/multisite/tasks/test_multi.yaml:27]: turns out the -a option is provided by the `nose.plugins.attrib` plugin, which wasn't being loaded by default. added a new commit to load the default plugins

**Comment by smanjara** [comment]
Do we still need 'rgw_bucket_shard_full_sync_marker' in rgw_bucket_shard_sync_info object which is now only tracking incremental sync?

**Comment by cbodley** [comment]
i pushed this to a feature branch at https://github.com/ceph/ceph/commits/wip-rgw-multisite-reshard so we can continue development there (and target PRs at it) without merging to master

**Pull request closed without merging**